docs/src/adr/020adr-arenas.md

```diff
@@ -0,0 +1,328 @@
+completely unnecessary, as we know for sure that the respective elements belong
+to the three sets. Moreover, when constructing `S \union T`, the rule `SetCupRule`
+creates five SMT constraints:
+
+```smt
+;; a, b, c belong to the union, when they belong to S
+(= in_a_u in_a_l)
+(= in_b_u in_b_l)
+(= in_c_u in_c_l)
+;; d and e belong to the union, when they belong to T
+(= in_d_u in_d_r)
+(= in_e_u in_e_r)
+```
+
+With the new representation, the set constructor would simply create five
+instances of `FixedElemPtr` that carry the value `true`, that is, the elements
+unconditionally belong to `S` and `T`. Further, the rule `SetCupRule` would
+simply copy the five pointers, without propagating anything to SMT.
+
+As a result, we obtain constant propagation of set membership, while
+keeping the general spirit of the arena-based encoding.
+
+### 3.3. Optimization 2: sharing membership in a powerset
+
+Consider the TLA+ operator that construct the powerset of `S`, that is, the set
+of all subsets of `S`:
+
+```tla
+SUBSET S
+```
+
+Let `c_S` be the cell that represents the set `S` and `c_1, ..., c_n` be the
+cells that represent the potential elements of `S`. Note that in general,
+membership of all these cells may be unknown in static. For example, consider
+the case when the set `S` is constructed from the following TLA+ expression:
+
+```tla
+{
+  x \in 1..100:
+    \E y \in 1..10:
+      y * y = x
+}
+```
+
+In the above example, computation of the predicate is delegated to the SMT
+solver.
+
+The code in `PowSetCtor` constructs 2^Cardinality(S) sets that contain all
+subsets of `S`. The tricky part here is that some of the elements of `S` may be
+outside of `S`. To deal with that, `PowSetCtor` constructs cells for each
+potential combinations of `c_1, ..., c_n` and adds membership tests for each of
+them. For instance, consider the subset `T` that is constructed by selecting the
+indices `1, 3, 5` of `1..n`. The constructor will introduce `3`
+constraints:
+
+```smt
+(= in_c_1_T in_c_1_S)
+(= in_c_3_T in_c_3_S)
+(= in_c_5_T in_c_5_S)
+```
+
+Hence, the current encoding introduces `2^n` SMT constants for the subsets and
+`n * 2^(n - 1)` membership constraints in SMT (thanks to Jure @Kukovec for
+telling me the precise formula).
+
+With the new representation, we would simply copy the respective membership
+pointers of the set `S`. This would require us to introduce zero constraints in
+the SMT, though we would still introduce `2^n` SMT constants to represent the
+subsets themselves. Note that we would still have to introduce `n * 2^(n - 1)`
+pointers in the arena. But this would be done during the process of rewriting.
+
+### 3.4. Feature: computing the set of functions via pointer sharing
+
+Sometimes, it happens that the model checker has to expand a set of functions
+`[S -> T]`. Such a set contains `|T|^|S|` functions. Since the model checker is
+working with arenas, it can only construct an arena representation of `[S ->
```

Comment:
```suggestion
`[S -> T]`. Such a set contains `|T|^|S|` functions. Since the model checker works with arenas, it can only construct an arena representation of `[S ->
```